dashboard/src/main/java/com/google/cloud/tools/opensource/cloudbomdashboard/VersionData.java

```diff
@@ -0,0 +1,277 @@
+      for (Artifact artifact : artifacts) {
+        String artifactId = artifact.getArtifactId();
+        //Concatenate this with the version of the current cloud BOM, so each entry has a unique
+        //key for the mapping in our table.
+        String artifactKey = artifactId + ":" + cloudBomVersion;
+        String groupId = artifact.getGroupId();
+        String version = artifact.getVersion();
+
+        String latestVersion = latestVersion(artifact);
+        String pomFileURL = getPomFileURL(groupId, artifactId, version);
+        String sharedDependencyVersion = sharedDependencyVersion(artifactKey, artifact,
+            sharedDependenciesPosition);
+
+        artifactIds.add(artifactId);
+        currentVersion.put(artifactKey, version);
+        newestVersion.put(artifactKey, latestVersion);
+        newestPomUrl.put(artifactKey, pomFileURL);
+        sharedDepsVersion.put(artifactKey, sharedDependencyVersion);
+        updatedTime.put(artifactKey, updatedTime(artifact));
+        metadataUrl.put(artifactKey, getMetadataUrl(artifact));
+      }
+    }
+
+    templateData.put("currentVersion", currentVersion);
+    templateData.put("sharedDependenciesPosition", sharedDependenciesPosition);
+    templateData.put("newestVersion", newestVersion);
+    templateData.put("newestPomUrl", newestPomUrl);
+    templateData.put("sharedDepsVersion", sharedDepsVersion);
+    templateData.put("updatedTime", updatedTime);
+    templateData.put("metadataUrl", metadataUrl);
+    templateData.put("artifacts", artifactIds);
+    templateData.put("versions", versions);
+    templateData.put("lastUpdated", LocalDateTime.now());
+    //Our all-versions dashboard page is a special case
+    if (versions.size() > 1) {
+      templateData.put("staticVersion", "All Versions");
+    } else if (versions.size() == 1) {
+      String onlyElement = versions.stream().findAny().get();
+      templateData.put("staticVersion", onlyElement);
+    }
+    return templateData;
+  }
+
+  private static String latestVersion(Artifact artifact) {
+    if (artifactToLatestVersion.containsKey(artifact)) {
+      return artifactToLatestVersion.get(artifact);
+    }
+    String metadataPath = getMetadataUrl(artifact);
+    File metadataFile = new File("metadata.xml");
+    try {
+      URL url = new URL(metadataPath);
+      FileUtils.copyURLToFile(url, metadataFile);
+      String latest = XmlGrabber.grabMetadataValue(metadataFile, "latest");
+      artifactToLatestVersion.put(artifact, latest);
+      return latest;
+    } catch (IOException ignored) {
+      artifactToLatestVersion.put(artifact, "");
+      return "";
+    }
+  }
+
+  private static String updatedTime(Artifact artifact) {
+    if (artifactToTime.containsKey(artifact)) {
+      return artifactToTime.get(artifact);
+    }
+    String metadataPath = getMetadataUrl(artifact);
+    File metadataFile = new File("metadata.xml");
+    try {
+      URL url = new URL(metadataPath);
```

Comment:
consider using org.apache.maven.artifact.repository.metadata.io.xpp3.MetadataXpp3Reader for this, possibly in a later PR, to handle some of the Maven variable resolution. There might also be a way to pull this from the local repo instead of hitting Maven central for every request